src/array.rs

```diff
@@ -730,10 +726,27 @@ impl<T: Element> PyArray<T, Ix1> {
     /// });
     /// ```
     pub fn from_slice<'py>(py: Python<'py>, slice: &[T]) -> &'py Self {
-        let array = PyArray::new(py, [slice.len()], false);
-        unsafe {
-            array.copy_ptr(slice.as_ptr(), slice.len());
+        let array = Self::new(py, [slice.len()], false);
+
+        // all other data types are assumed to be `Copy` by NumPy
+        if T::DATA_TYPE == DataType::Object {
+            // keep array referenced as long as its contents is not initialized
+            let ref_ = mem::ManuallyDrop::new(array.to_object(py));
```

Comment:
To avoid leaking memory if clone panics, you could use a guard pattern similar to https://github.com/PyO3/pyo3/blob/00c84eb0baec6f41623e83737a291d3e0d30cc5b/src/conversions/array.rs#L93

You would need to drop the initialized elements by hand and then zero the array so that numpy can deallocate safely I guess.

Reply:
Certainly, however I prefer to postpone this to separate PR after we have reached soundness w.r.t `PyArray` creation.

Reply:
👍 makes sense, let's just remember to open an issue for this when this PR merges.

Reply:
I took a different approach for now:
```rust
// Use zero-initialized pointers for object arrays
// so that partially initialized arrays can be dropped safely
// in case the iterator implementation panics.
let array = if T::DATA_TYPE == DataType::Object {
    Self::zeros(py, [iter.len()], false)
} else {
    Self::new(py, [iter.len()], false)
};
```
This way we start out with null pointers in the object case and the array is always safe to drop. I think the overhead is warranted due to the cost that arrays of pointers imply in any case. (Of course, using a guard that just zeros out  uninitialized array elements if there actually is a panic can still be implemented as a follow-up optimization.)

Reply:
Oh very nice!

Reply:
Yeah, it looks a good temporal solution 👍🏼